### PR TITLE
Feed forward chunking

### PR DESCRIPTION
**Pull request by Pradhy729**
Official PR for #5928

### CODE REVIEW

**Comment by Pradhy729** [comment]
@patrickvonplaten - here's an initial implementation I have. My first step is to get the model to work with chunked feed forward - and it works! I still need to run the benchmark test to find out the benefits in terms of memory.

However, I see a problem. The new architecture causes some of the nn.Module weights and bias parameter-names to change - which would be a problem with loading existing pretrained weights from checkpoints. 
For example:
`bert.encoder.layer.0.intermediate.dense.weight` --> becomes `bert.encoder.layer.0.feed_forward.dense.dense.weight`

See the failing tests for more details. Any thoughts/ideas on how to get around this?

File: src/transformers/configuration_utils.py [hunk 1]
```diff
@@ -159,6 +159,7 @@ def __init__(self, **kwargs):
         self.no_repeat_ngram_size = kwargs.pop("no_repeat_ngram_size", 0)
         self.bad_words_ids = kwargs.pop("bad_words_ids", None)
         self.num_return_sequences = kwargs.pop("num_return_sequences", 1)
+        self.chunk_size_feed_forward = kwargs.pop("chunk_size_feed_forward", 1)
```
Comment (patrickvonplaten) [inline src/transformers/configuration_utils.py:162]: This should be disabled by default -> so would be nice to set it to `0`

Comment (patrickvonplaten) [reply src/transformers/configuration_utils.py:162]: And to make sure we don't break backward compatibility...

File: src/transformers/modeling_bert.py [hunk 2]
```diff
@@ -371,15 +376,34 @@ def forward(self, hidden_states, input_tensor):
         return hidden_states
 
 
+class ChunkFeedForward(nn.Module):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_bert.py:379]: please rename to `ChunkBertFeedForward`

Comment (patrickvonplaten) [reply src/transformers/modeling_bert.py:379]: Actually, as you said this won't be possible because the layer weights would change, so we have to keep `self.intermediate` and `self.output` sadly...

File: src/transformers/modeling_bert.py [hunk 3]
```diff
@@ -408,8 +432,7 @@ def forward(
             attention_output = cross_attention_outputs[0]
             outputs = outputs + cross_attention_outputs[1:]  # add cross attentions if we output attention weights
 
-        intermediate_output = self.intermediate(attention_output)
```
Comment (patrickvonplaten) [inline src/transformers/modeling_bert.py:411]: To solve the problem, my suggestions would be to wrap these two calls in a function `forward_chunk` which is part of this class (`def forward_chunk(self, ....)`) and call `apply_chunking_to_forward(self.chunk_size_feed_forward, self.seq_len_dim, self.forward_chunk, attention_output,)` here.

Comment (Pradhy729) [reply src/transformers/modeling_bert.py:411]: Don't think I quite follow what you mean here. Which two calls do you want to wrap?
Did you mean to have a `forward_chunk` function in the `BertLayer` class?

Comment (Pradhy729) [reply src/transformers/modeling_bert.py:413]: Ok I fixed it based on your input - looks ok to me now.

Comment (patrickvonplaten) [reply src/transformers/modeling_bert.py:413]: Great, that's exactly what I meant :-)

File: tests/test_modeling_reformer.py [hunk 4]
```diff
@@ -577,6 +555,7 @@ class ReformerLocalAttnModelTest(ReformerTesterMixin, ModelTesterMixin, unittest
     test_pruning = False
     test_headmasking = False
     test_torchscript = False
+    test_feed_forward_chunking = True
```
Comment (patrickvonplaten) [inline tests/test_modeling_reformer.py:558]:
```suggestion
    test_chunking = True
```

**Comment by hamad8395** [comment]
hh

File: tests/test_modeling_reformer.py (hunk 4 above)
Comment (hamad8395) [reply tests/test_modeling_reformer.py]: hh

Comment (hamad8395) [reply tests/test_modeling_reformer.py]: hh

**Comment by patrickvonplaten** [comment]
Hey @Pradhy729, thanks a lot for continuing the PR. I made a couple of changes: fix the docs and added tests for all models, whereas only Reformer and Bert tests are on for now. 

 Would be great if @LysandreJik @sgugger @thomwolf @sshleifer can review. 

This PR shows how `feed_forward_chunking` can be employed for all models. Feed forward chunking is explained here: https://huggingface.co/blog/reformer#2-chunked-feed-forward-layers in combination with some benchmarking. It can give good memory improvements for certain model architectures.
For Bert a test is added showing that the model gives equal results. This function can easily be added to other models, the same way it was done for BERT. There is no real drawback in implementing this IMO. 

**To-Do after review is positive**:
1. Add feed forward chunking to more models. @Pradhy729, feel free to add it to as many models as you want. The rest can also be added in a new PR or we open a "first good issue" for it.
2. Add feed forward chunking for language modeling loss. Chunking of feed forward layers in the attention block is often not really helpful to save memory - only if the model has very few attention heads. On the other hand, a real bottleneck is often the last word embedding layer. When training the loss does not have to be calculated in one huge batch (over time dim), but can be chunked the same way it is done here for Feed forward layers. This is not even really implemented in Reformer yet and would definitely require a new PR.

**Comment by Pradhy729** [comment]
Great! Thanks @patrickvonplaten 
Will wait for reviewers and start working on the others.

**Review by sgugger — APPROVED** [review]
This is very useful work, thanks for tackling this!

File: src/transformers/configuration_utils.py [hunk 5]
```diff
@@ -64,6 +64,11 @@ class PretrainedConfig(object):
                 2.
             xla_device (:obj:`bool`, `optional`):
                 A flag to indicate if TPU are available or not.
+            chunk_size_feed_forward (:obj:`int`, `optional`, defaults to :obj:`0`):
```
Comment (patrickvonplaten) [inline src/transformers/configuration_utils.py:67]:
```suggestion
            chunk_size_feed_forward (:obj:`int`, `optional`, defaults to 0):
```

File: src/transformers/modeling_bert.py [hunk 6]
```diff
@@ -410,11 +418,17 @@ def forward(
...
-        layer_output = self.output(intermediate_output, attention_output)
+        layer_output = apply_chunking_to_forward(
+            self.chunk_size_feed_forward, self.seq_len_dim, self.feed_forward_chunk, attention_output
+        )
```
Comment (LysandreJik) [inline src/transformers/modeling_bert.py:423]: Very much a nitpick here, for future PRs probably, but this looks a lot like the gradient checkpointing method from PyTorch. This method takes the callable (the forward) method as first positional argument and I think it makes sense to have it this way.

**Review by LysandreJik — APPROVED** [review]
Very nice, love how it only requires a few lines of code now that `apply_chunking_to_forward` is created.

File: src/transformers/modeling_bert.py (hunk 6 above)
Comment (Pradhy729) [reply src/transformers/modeling_bert.py:423]: I can do this globally in the new PR where I add the chunking for other models. Let me know if you have concerns with that.

File: src/transformers/configuration_utils.py (hunk 1 above)
Comment (Franciszzj) [reply src/transformers/configuration_utils.py]: in modeling_bert.py, refer to #6972. thanks.
